File: client/src/lib/redact.ts
// redact.ts — strips credentials out of text before it can reach a
// bug report (#721).
//
// The in-app report inlines the client log in a GitHub issue body
// (ADR 0017 §7). That log once carried the WebSocket URL verbatim,
// `?token=` and all, so every report with a connect line published
// the reporter's session token. Durable sessions (#517) would make
// such a token outlive a server restart, which is why this landed
// first.
//
// Two layers use this module:
//
//   - at the source: ws.ts logs redactURL(url), never the raw URL;
//   - as a backstop: the ws log buffer, the console/error capture, the
//     merged bug-report log and the submitted draft all pass through
//     redactSecrets, so a log line added tomorrow can't reintroduce
//     the leak.
//
// The server applies the same rules again before rendering the issue
// (server/internal/util/redact), so a stale client can't leak either.
// Keep the two rule sets in step.
//
// Pure and dependency-free: vitest covers it directly.

export const REDACTED = "REDACTED";

// SECRET_KEY matches a key that names a credential anywhere it appears
// as key=value: `token`, `access_token`, `refresh_token`, `id_token`,
// `client_secret`, `password`, the reclaim `ticket`, … A suffix match
// rather than a list, so `csrf_token` or `session_token` are covered
// without anyone remembering to add them.
const SECRET_KEY = String.raw`[a-z0-9_-]*(?:token|secret|password|passwd|ticket)`;

// QUERY_ONLY_KEY is a credential key too short to trust in free text.
// `t` is the invite and reclaim token (`#/games/<id>/join?t=…`); it is
// only redacted when it is unambiguously a query parameter, so a log
// line saying "t=3" about a turn is left alone.
const QUERY_ONLY_KEY = "t";

// VALUE is one key's value: everything up to the next parameter or
// fragment separator, whitespace, quote or bracket. Percent-escapes are
// consumed as part of the value EXCEPT %23 (#) and %26 (&), which end
// it — that is what keeps a URL-encoded URL (`…%3Ftoken%3Dabc%26game%3D1`)
// from losing its trailing parameters to the redaction.
const VALUE = String.raw`(?:[^&#\s"'<>()\[\]{},;%\\]|%(?:[013-9a-f][0-9a-f]|2[0-2457-9a-f]))+`;

// EQ is a literal or percent-encoded equals sign.
const EQ = "(=|%3d)";

// Key=value anywhere: the key must not be glued to a preceding word
// character, so "bearer_tokens=3" style prose is the worst false
// positive and "notatoken=x" is not one.
const KEY_VALUE_RE = new RegExp(String.raw`(^|[^a-z0-9_-])(${SECRET_KEY})${EQ}(${VALUE})`, "gi");

// `?t=` / `&t=` (and their encoded forms `%3Ft=`, `%26t%3D`, and the
// HTML-escaped `&amp;t=`).
const QUERY_ONLY_RE = new RegExp(
  String.raw`(\?|&amp;|&|%3f|%26)(${QUERY_ONLY_KEY})${EQ}(${VALUE})`,
  "gi",
);

// JSON-shaped: "token":"abc", including the escaped form a stringified
// object picks up inside another string (\"token\":\"abc\").
const JSON_RE = new RegExp(
  String.raw`(\\?")(${SECRET_KEY})(\\?"\s*:\s*\\?")((?:[^"\\]|\\(?!"))*)(\\?")`,
  "gi",
);

// `Authorization: Bearer <credential>`, `Bearer%20<credential>`. Case
// sensitive and length-gated on purpose: "Bearer of the Heavens" is a
// card, and a real credential is never four letters long. The guard
// before the word is "not a letter" rather than \b so the encoded
// `%20Bearer%20…` form (where a digit touches the B) still matches.
const BEARER_RE = /(^|[^A-Za-z])(Bearer)(\s+|%20|\+)([A-Za-z0-9._~+/=-]{16,})/g;

// redactSecrets replaces every credential value it recognises in s with
// REDACTED, keeping the key so a triager can still see that one was
// sent. Idempotent: redacting a redacted string changes nothing.
export function redactSecrets(s: string): string {
  if (!s) return s;
  return s
    .replace(
      KEY_VALUE_RE,
      (_m, pre: string, key: string, eq: string) => `${pre}${key}${eq}${REDACTED}`,
    )
    .replace(
      QUERY_ONLY_RE,
      (_m, pre: string, key: string, eq: string) => `${pre}${key}${eq}${REDACTED}`,
    )
    .replace(
      JSON_RE,
      (_m, open: string, key: string, mid: string, _v: string, close: string) =>
        `${open}${key}${mid}${REDACTED}${close}`,
    )
    .replace(
      BEARER_RE,
      (_m, pre: string, word: string, sep: string) => `${pre}${word}${sep}${REDACTED}`,
    );
}

// redactURL renders a URL safe to log: scheme, host, path, fragment and
// the non-secret parameters survive (`game` and `player` are what a
// triager needs to find the table); credential values become REDACTED.
//
// Deliberately string-based rather than URL/URLSearchParams: those
// re-encode the parameters, so "REDACTED" would be the only thing in
// the line that looked like the original, and a relative or malformed
// URL would throw where a log line must never throw.
export function redactURL(url: string): string {
  return redactSecrets(url);
}

File: server/internal/util/redact/redact.go
// Package redact strips credentials out of text before it is published
// anywhere a session token must not appear — today, the body of an
// in-app bug report filed as a GitHub issue (#721, ADR 0017 §9).
//
// Client-supplied text is the risk. The browser's log once carried the
// WebSocket URL verbatim, `?token=` and all, and the bug-report renderer
// inlined it, so every report with a connect line published the
// reporter's session token. The client now redacts at the source
// (client/src/lib/redact.ts); this package is the server's copy of the
// same rules, so an old tab, a curl, or a log line nobody thought about
// cannot leak through the server either. Keep the two rule sets in step.
package redact

import "regexp"

// Redacted replaces a credential value. Keys are kept, so a triager can
// still see that a token was sent without being able to use it.
const Redacted = "REDACTED"

const (
	// secretKey matches a key that names a credential: token,
	// access_token, refresh_token, id_token, client_secret, password,
	// the reclaim ticket, … A suffix match rather than a list, so a new
	// `*_token` is covered without anyone remembering to add it.
	secretKey = `[a-z0-9_-]*(?:token|secret|password|passwd|ticket)`

	// value is one key's value: everything up to the next parameter or
	// fragment separator, whitespace, quote or bracket. Percent-escapes
	// are part of the value EXCEPT %23 (#) and %26 (&) — and their
	// double-encoded %2523 / %2526 — which end it. That keeps a
	// URL-encoded URL (`…%3Ftoken%3Dabc%26game%3D1`) from losing its
	// trailing parameters to the redaction.
	//
	// The escape branches: a single escape other than %23/%25/%26; a
	// double escape (%25XX) other than %2523/%2526.
	value = `(?:[^&#\s"'<>()\[\]{},;%\\]|%(?:[013-9a-f][0-9a-f]|2[0-2479a-f])|%25(?:[013-9a-f][0-9a-f]|2[0-2457-9a-f]))+`

	// eq is a literal, percent-encoded or double-encoded equals sign
	// (`=`, `%3D`, `%253D` — the last is a URL carried inside an encoded
	// parameter of another URL).
	eq = `(=|%3d|%253d)`
)

var (
	// key=value anywhere, the key not glued to a preceding word
	// character ("notatoken=x" is left alone).
	keyValueRE = regexp.MustCompile(`(?i)(^|[^a-z0-9_-])(` + secretKey + `)` + eq + value)

	// `t` is the invite and reclaim token (`#/games/<id>/join?t=…`). Too
	// short to trust in prose ("at t=3"), so only redacted as a query
	// parameter: after ?, &, &amp;, or their encoded and double-encoded
	// forms.
	queryOnlyRE = regexp.MustCompile(`(?i)(\?|&amp;|&|%253f|%2526|%3f|%26)(t)` + eq + value)

	// JSON-shaped: "token":"abc", and the escaped form a stringified
	// object picks up inside another string (\"token\":\"abc\").
	jsonRE = regexp.MustCompile(`(?i)(\\?")(` + secretKey + `)(\\?"\s*:\s*\\?")(?:[^"\\]|\\[^"])*(\\?")`)

	// `Authorization: Bearer <credential>`. Case-sensitive and
	// length-gated on purpose: "Bearer of the Heavens" is a card name,
	// and a real credential is never four letters long.
	bearerRE = regexp.MustCompile(`(^|[^A-Za-z])(Bearer)(\s+|%20|\+)[A-Za-z0-9._~+/=-]{16,}`)
)

// Secrets returns s with every credential value it recognises replaced
// by Redacted: token-bearing query parameters (literal and URL-encoded),
// the invite `t`, JSON token fields, and Bearer credentials. Idempotent.
func Secrets(s string) string {
	if s == "" {
		return s
	}
	s = keyValueRE.ReplaceAllString(s, "${1}${2}${3}"+Redacted)
	s = queryOnlyRE.ReplaceAllString(s, "${1}${2}${3}"+Redacted)
	s = jsonRE.ReplaceAllString(s, "${1}${2}${3}"+Redacted+"${4}")
	s = bearerRE.ReplaceAllString(s, "${1}${2}${3}"+Redacted)
	return s
}
